### PR TITLE
Client add option set

### PR DESCRIPTION
**Pull request by outscale-nts**

### CODE REVIEW

File: api/client/sg.cc [hunk 1]
```diff
@@ -481,7 +491,63 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        sub_sg_rule_del_help();
+        return 1;
+    }
+
+    string sg = string(argv[4]);
+    if (is_hash(string(argv[5])) == true) {
+        return sub_sg_hashed_rule_del(sg, string(argv[5]), options);
+    } else {
+        return sub_sg_param_rule_del(argc, argv, options);
+    }
```
Comment (outscale-toa) [inline api/client/sg.cc]: This function must return a value when no condition met

File: api/client/nic.cc [hunk 2]
```diff
@@ -271,6 +271,37 @@ static int sub_nic_sg_add(int argc, char **argv,
...
+    }
+
+    if (argc <= 5) {
+        sub_nic_sg_set_help();
+        return 1;
+    }
+    string nic = string(argv[4]);
+    vector<string> to_set;
+    for (int i = 5; i < argc; i++) {
+        string a = string(argv[i]);
```
Comment (outscale-toa) [inline api/client/nic.cc]: can you rename this variable "a" please ?

File: api/client/nic.cc [hunk 3]
```diff
@@ -271,6 +271,37 @@ static int sub_nic_sg_add(int argc, char **argv,
...
+        if (a[0] == '-')
+            break;
+        to_set.push_back(a);
+    }
+    if (!to_set.size()) {
+        to_set.push_back("");
+    }
+
```
Comment (outscale-toa) [inline api/client/nic.cc]: can you remove blank line please?

File: api/client/sg.cc [hunk 4]
```diff
@@ -421,21 +421,31 @@ static void sub_sg_rule_del_help(void) {
...
+        isHex = false;
+
+        for (int j = 0; j < hexLength ; j++) {
+            if (param[i] == hexValues[j]) {
+                isHex = true;
+                break;
+            }
+        }
+        if (!isHex)
+            break;
```
Comment (outscale-toa) [inline api/client/sg.cc]: I think there are another way to do this without using two for loops

File: api/client/nic.cc (hunk 2 above)
Comment (outscale-toa) [inline api/client/nic.cc]: can you rename this variable "a" please ?

File: tests/functions.sh [hunk 5]
```diff
@@ -1123,27 +1123,8 @@ function sg_rule_del_icmp {
...
-            mask_size: 0
-          }
-        }
-      }
-    }
-  }
-}
-" > $f
-    request $but_id $f
+
```
Comment (outscale-mgo) [inline tests/functions.sh]: useless space

File: tests/functions.sh [hunk 6]
```diff
@@ -1139,27 +1139,8 @@ function sg_rule_del_icmp6 {
...
```
Comment (outscale-mgo) [inline tests/functions.sh]: space

File: api/client/sg.cc [hunk 7]
```diff
@@ -421,21 +421,13 @@ static void sub_sg_rule_del_help(void) {
...
-    }
-
-    if (argc <= 5) {
-        sub_sg_rule_del_help();
-        return 1;
-    }
-
-    string sg = string(argv[4]);
-    string hash = string(argv[5]);
+static bool is_hash(string param) {
```
Comment (outscale-mgo) [inline api/client/sg.cc]: you should add a space before the function

File: api/client/sg.cc [hunk 8]
```diff
@@ -421,21 +421,13 @@ static void sub_sg_rule_del_help(void) {
...
+    if (param.find("-") == 0)
+        return false;
+    return true;
+}
```
Comment (outscale-mgo) [inline api/client/sg.cc]: here too

File: api/client/sg.cc [hunk 9]
```diff
@@ -421,21 +421,13 @@ static void sub_sg_rule_del_help(void) {
...
-
-    if (argc <= 5) {
-        sub_sg_rule_del_help();
-        return 1;
-    }
-
-    string sg = string(argv[4]);
-    string hash = string(argv[5]);
+static bool is_hash(string param) {
+    if (param.find("-") == 0)
```
Comment (outscale-mgo) [inline api/client/sg.cc]: return !param.find("-")

File: api/client/sg.cc [hunk 10]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
 
     proto::Messages delete_res;
     return request(delete_req, &delete_res, options, false) ||
-        check_request_result(delete_res);
+            check_request_result(delete_res);
+}
```
Comment (outscale-mgo) [inline api/client/sg.cc]: here too

File: api/client/sg.cc [hunk 11]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+            "      }"
+            "    }"
+            "  }"
+            "}";
+
+    proto::Messages res;
+    if (request(req, &res, options, false) || check_request_result(res))
+        return 1;
+    return 0;
+}
```
Comment (outscale-mgo) [inline api/client/sg.cc]: here too

File: api/client/sg.cc [hunk 12]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
 
     proto::Messages delete_res;
     return request(delete_req, &delete_res, options, false) ||
-        check_request_result(delete_res);
+            check_request_result(delete_res);
+}
+static int sub_sg_param_rule_del(int argc, char **argv,
```
Comment (outscale-mgo) [inline api/client/sg.cc]: s/int/bool/

File: api/client/sg.cc [hunk 13]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+                                 const GlobalOptions &options) {
+    RuleAddOptions opts;
+    if (opts.parse(argc, argv)) {
+        sub_sg_rule_del_help();
+        return 1;
```
Comment (outscale-mgo) [inline api/client/sg.cc]: true;

File: api/client/sg.cc [hunk 14]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+    req +=
+            "        }"
+            "      }"
+            "    }"
+            "  }"
+            "}";
+
+    proto::Messages res;
+    if (request(req, &res, options, false) || check_request_result(res))
+        return 1;
```
Comment (outscale-mgo) [inline api/client/sg.cc]: s/1/true/

File: api/client/sg.cc [hunk 15]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+    return 0;
```
Comment (outscale-mgo) [inline api/client/sg.cc]: s/0/false/

File: api/client/sg.cc [hunk 16]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        "messages {"
+        "  revision: " PROTO_REV
+        "  message_0 {"
+        "    request {"
+        "      sg_rule_del {"
+        "        sg_id: \"" + opts.sg + "\""
+        "        rule {"
+        "          direction: " + opts.direction +
+        "          protocol: " + to_string(opts.proto);
+    if (opts.has_port_start)
```
Comment (outscale-mgo) [inline api/client/sg.cc]: add brace on this if, as the body is 2 lines long

File: api/client/sg.cc [hunk 17]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+            "}";
+
+    proto::Messages res;
+    if (request(req, &res, options, false) || check_request_result(res))
+        return 1;
+    return 0;
+}
+static int sub_sg_rule_del(int argc, char **argv,
+                            const GlobalOptions &options) {
+    if (argc >= 5 && string(argv[4]) == "help") {
```
Comment (outscale-mgo) [inline api/client/sg.cc]: use strcmp here

File: api/client/sg.cc [hunk 18]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        sub_sg_rule_add_help();
+        return 0;
+    }
+
+    if (argc <= 5) {
+        sub_sg_rule_del_help();
+        return 1;
+    }
+
+    string sg = string(argv[4]);
```
Comment (outscale-mgo) [inline api/client/sg.cc]: why do you use string here ? is there any operation here that you could not have done with "char *" ?

File: api/client/sg.cc [hunk 19]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+    if (is_hash(string(argv[5])) == true) {
```
Comment (outscale-mgo) [inline api/client/sg.cc]: same here, I don't understand why you are constructing a string

File: api/client/sg.cc [hunk 20]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        return sub_sg_hashed_rule_del(sg, string(argv[5]), options);
```
Comment (outscale-mgo) [inline api/client/sg.cc]: same here

**Comment by outscale-mgo** [comment]
really really need a rebase on origin/next

File: api/client/sg.cc [hunk 21]
```diff
@@ -481,7 +473,62 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+            req += "   security_group: \"" + opts.sg_members + "\"";
+    req +=
+            "        }"
+            "      }"
+            "    }"
+            "  }"
+            "}";
+
+    proto::Messages res;
+    if (request(req, &res, options, false) || check_request_result(res))
```
Comment (outscale-mgo) [inline api/client/sg.cc]: `return request(req, &res, options, false) || check_request_result(res);`

**Pull request closed without merging**